File: day4/main.ts
import { readFileSync } from "fs";

type Point = { r: number; c: number };

const dirs: Point[] = [
  { r: -1, c: -1 },
  { r: -1, c: 0 },
  { r: -1, c: 1 },
  { r: 0, c: -1 },
  { r: 0, c: 1 },
  { r: 1, c: -1 },
  { r: 1, c: 0 },
  { r: 1, c: 1 },
];

function key(p: Point): string {
  return `${p.r},${p.c}`;
}

function parseGrid(lines: string[]): Set<string> {
  const rolls = new Set<string>();
  lines.forEach((line, r) => {
    for (let c = 0; c < line.length; c++) {
      if (line[c] === "@") {
        rolls.add(key({ r, c }));
      }
    }
  });
  return rolls;
}

function neighborCounts(rolls: Set<string>): Map<string, number> {
  const counts = new Map<string, number>();
  rolls.forEach((k) => {
    const [r, c] = k.split(",").map(Number);
    let cnt = 0;
    for (const d of dirs) {
      if (rolls.has(key({ r: r + d.r, c: c + d.c }))) cnt++;
    }
    counts.set(k, cnt);
  });
  return counts;
}

function part1(counts: Map<string, number>): number {
  let acc = 0;
  counts.forEach((v) => {
    if (v < 4) acc++;
  });
  return acc;
}

function part2(rolls: Set<string>, counts: Map<string, number>): number {
  const queue: string[] = [];
  const inQ = new Set<string>();
  counts.forEach((v, k) => {
    if (v < 4) {
      queue.push(k);
      inQ.add(k);
    }
  });
  let removed = 0;
  while (queue.length) {
    const k = queue.shift()!;
    if (!rolls.has(k)) continue;
    rolls.delete(k);
    removed++;
    const [r, c] = k.split(",").map(Number);
    for (const d of dirs) {
      const nk = key({ r: r + d.r, c: c + d.c });
      if (rolls.has(nk)) {
        const newCnt = (counts.get(nk) || 0) - 1;
        counts.set(nk, newCnt);
        if (newCnt < 4 && !inQ.has(nk)) {
          queue.push(nk);
          inQ.add(nk);
        }
      }
    }
  }
  return removed;
}

const lines = readFileSync("input.txt", "utf8").trim().split(/\r?\n/);
const rolls = parseGrid(lines);
const counts = neighborCounts(rolls);
const t0 = performance.now();
const a = part1(counts);
const removed = part2(new Set(rolls), new Map(counts));
const elapsed = performance.now() - t0;
console.log(
  `accessible=${a} removable_total=${removed} elapsed_ms=${elapsed.toFixed(3)}`,
);

File: day4/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"time"
)

type Point struct{ r, c int }

var dirs = []Point{
	{-1, -1}, {-1, 0}, {-1, 1},
	{0, -1}, {0, 1},
	{1, -1}, {1, 0}, {1, 1},
}

func parseGrid(lines []string) map[Point]bool {
	rolls := make(map[Point]bool)
	for r, line := range lines {
		for c, ch := range line {
			if ch == '@' {
				rolls[Point{r, c}] = true
			}
		}
	}
	return rolls
}

func neighborCounts(rolls map[Point]bool) map[Point]int {
	counts := make(map[Point]int)
	for p := range rolls {
		cnt := 0
		for _, d := range dirs {
			if rolls[Point{p.r + d.r, p.c + d.c}] {
				cnt++
			}
		}
		counts[p] = cnt
	}
	return counts
}

func part1(counts map[Point]int) int {
	acc := 0
	for _, v := range counts {
		if v < 4 {
			acc++
		}
	}
	return acc
}

func part2(rolls map[Point]bool, counts map[Point]int) int {
	type queueItem Point
	q := make([]queueItem, 0)
	inQ := make(map[Point]bool)
	for p, c := range counts {
		if c < 4 {
			q = append(q, queueItem(p))
			inQ[p] = true
		}
	}
	removed := 0
	for len(q) > 0 {
		p := q[0]
		q = q[1:]
		pt := Point(p)
		if !rolls[pt] {
			continue
		}
		delete(rolls, pt)
		removed++
		for _, d := range dirs {
			np := Point{pt.r + d.r, pt.c + d.c}
			if _, ok := rolls[np]; ok {
				counts[np]--
				if counts[np] < 4 && !inQ[np] {
					q = append(q, queueItem(np))
					inQ[np] = true
				}
			}
		}
	}
	return removed
}

func main() {
	file, _ := os.Open("input.txt")
	defer file.Close()
	var lines []string
	sc := bufio.NewScanner(file)
	for sc.Scan() {
		lines = append(lines, sc.Text())
	}

	rolls := parseGrid(lines)
	counts := neighborCounts(rolls)

	start := time.Now()
	a := part1(counts)
	removed := part2(rolls, counts)
	elapsed := time.Since(start).Seconds() * 1000
	fmt.Printf("accessible=%d removable_total=%d elapsed_ms=%.3f\n", a, removed, elapsed)
}
